File: 总题库/990.SatisfiabilityofEqualityEquations.py
class UnionFind:

    def __init__(self):
        self.parents = [i for i in range(26)]
        self.rank = [0 for _ in range(26)]

    def find(self, index: int) -> int:
        if index == self.parents[index]:
            return index

        self.parents[index] = self.find(self.parents[index])
        return self.parents[index]

    def unite(self, index1: int, index2: int) -> int:
        root1 = self.find(index1)
        root2 = self.find(index2)

        if root1 != root2:
            if self.rank[root1] < self.rank[root2]:
                self.parents[root1] = root2
            elif self.rank[root1] > self.rank[root2]:
                self.parents[root2] = root1
            else:
                self.parents[root2] = root1
                self.rank[root1] += 1


class Solution:
    def equationsPossible(self, equations: List[str]) -> bool:
        uf = UnionFind()
        for s in equations:
            if s[1] == '=':
                index1 = ord(s[0]) - ord('a')
                index2 = ord(s[3]) - ord('a')
                uf.unite(index1, index2)

        for s in equations:
            if s[1] == '!':
                index1 = ord(s[0]) - ord('a')
                index2 = ord(s[3]) - ord('a')
                if uf.find(index1) == uf.find(index2):
                    return False

        return True

File: 总题库/990.SatisfiabilityofEqualityEquations.cpp
// 并查集
class UnionFind {
private:
    vector<int> parent;
    vector<int> rank;
public:
    UnionFind() {
        parent.resize(26);
        rank.resize(26);
        // iota函数对一个范围数据进行赋值, 从第三个参数，依此往后赋值；
        iota(parent.begin(), parent.end(), 0);
        iota(rank.begin(), rank.end(), 0);
    }

    // 并查集的核心步骤，寻找一个点所属连通集的根节点；
    int find(int index) { // 找index所属连通集的根节点；
        if (index == parent[index]) {
            // 此时说明index是它当前所属连通集的根节点；
            return index;
        }
        parent[index] = find(parent[index]);
        return parent[index];
    }

    // 将index1和index2连通；
    void unite(int index1, int index2) {
        int root1 = find(index1);
        int root2 = find(index2);
        
        if (root1 != root2) {
            if (rank[root1] < rank[root2]) {
                parent[root1] = root2;
            }
            else if (rank[root1] > rank[root2]) {
                parent[root2] = root1;
            }
            else {
                parent[root2] = root1;
                rank[root1]++;
            }
        }
    }
};

class Solution {
public:
    bool equationsPossible(vector<string>& equations) {
        UnionFind uf; // 先构造一个空的连通集
        for (const string & str : equations) {
            if (str[1] == '=') {
                // 先将所有等式构造成连通集；
                int index1 = str[0] - 'a';
                int index2 = str[3] - 'a';
                uf.unite(index1, index2);
            }
        }
        for (const string & str : equations) {
            if (str[1] == '!') {
                int index1 = str[0] - 'a';
                int index2 = str[3] - 'a';
                if (uf.find(index1) == uf.find(index2)) {
                    return false;
                }
            }
        }
        return true;
    }
};